src/storage/mvcc/reader/forward_scanner.rs

```diff
@@ -180,83 +200,178 @@ impl<S: Snapshot> ForwardScanner<S> {
-                            (Key::from_encoded(wk.to_vec()).truncate_ts()?, true, false)
+                            // Write cursor user key < lock cursor, it means the lock of the
+                            // current key that write cursor is pointing to does not exist.
+                            current_user_key_slice = Key::truncate_ts_for(wk)?;
+                            has_write = true;
+                            has_lock = false;
+                        }
+                        Ordering::Greater => {
+                            // Write cursor user key > lock cursor, it means we got a lock of a
+                            // key that does not have a write. In SI, we need to check if the
+                            // lock will cause conflict.
+                            current_user_key_slice = lk;
+                            has_write = false;
+                            has_lock = true;
+                        }
+                        Ordering::Equal => {
+                            // Write cursor user key == lock cursor, it means the lock of the
+                            // current key that write cursor is pointing to *exists*.
+                            current_user_key_slice = lk;
+                            has_write = true;
+                            has_lock = true;
                         }
-                        Ordering::Greater => (Key::from_encoded(lk.to_vec()), false, true),
-                        Ordering::Equal => (Key::from_encoded(lk.to_vec()), true, true),
                     },
-                }
-            };
+                };
+                // Convert the key into a `Vec` to avoid data being invalidated after cursor
+                // moving.
+                current_user_key = Key::from_encoded(current_user_key_slice.to_vec());
+            }
 
-            let lock = if has_lock {
-                Some(self.lock_cursor.value(&mut self.statistics.lock).to_vec())
-            } else {
-                None
-            };
-            let res = self.get(&key, lock);
+            // Attempt to read specified version of the key. Note that we may get `None`
+            // indicating that no desired version is found, or a DELETE version is found, or
+            // when `has_write == false`. We may also get `Err` due to key lock. We need to
+            // ensure that the scanner keeps working for future calls even when meeting key
+            // lock errors (why??), so we don't apply `?` operator to the result here thus
+            // iterators can be moved forward.
+            let result = self.get(&current_user_key, has_write, has_lock);
 
             if has_write {
-                let next_seek_key = key.clone().append_ts(0);
-                self.write_cursor
-                    .near_seek(&next_seek_key, false, &mut self.statistics.write)?;
+                // After `self.get()`, our write cursor may be pointing to current user key (if we
```

Comment:
You can extract following code into a function like `move_iterator_to_next_row`.